handbook/engineering/introducing_a_new_service.md

```diff
@@ -0,0 +1,45 @@
+# Introducing a new service
+
+Before reading this document be sure to first check out our [architecture overview](https://docs.sourcegraph.com/dev/background-information/architecture).
+
+## Terminology
+
+When we say "service" here we are referring to code that runs _within a Docker container_. This may be one or more processes, servers, goroutines, threads, etc. operating in a single Docker container.
+
+## When does introducing a new service make sense?
+
+Sourcegraph is composed of several smaller services (gitserver, searcher, symbols, etc.) and a single monolithic service (the frontend).
+
+When thinking of adding a new service, it is important to think through the following questions carefully:
+
+1. Does the code belong in an existing service?
+2. Instead of introducing a new service, could the process/etc reasonably live inside the frontend monolith as a singleton background worker, goroutine, etc.?
+3. Does the code rely very heavily on APIs that exist (or will need to be introduced) in another service? If so, have you considered putting it in that service instead?
+4. If you make the change within an existing service, instead of a new one, would it substantially increase the complexity of the task at hand?
+    - For example, the service you are writing _must_ be written in language X and it is impossible/very difficult to integrate language X into one of our existing Go services.
+5. Does it need its own resource constraints and scaling?
+   - For example, the service you are creating needs its own CPU / memory resource constraints, or must be able to scale horizontally across machines.
+
+If after asking the above questions to yourself you still believe introducing a new service is the best approach forward, you should [create an RFC](../communication/rfcs/index.md) proposing it to the rest of the team. In your RFC, be sure to answer the above questions and explaining why you think a seperate service is warranted.
+
+### Services have additional overhead for us and users that is easy to forget
+
+Introducing a new service means additional complexity and overhead for us and users. While they may provide a good way to isolate code, services are very much not free.
```

Comment:
Im not sure how to rephrase this (and it might be out of the scope here so do not let this be a blocker) but I think this implies that bundling it in the monolith is by extension free or at least "cheaper" which is not accurate.

Reply:
```suggestion
Introducing a new service means additional complexity and overhead for us and users. While they may provide a good way to isolate code, services are very much not free. Of course, a monolith is not free or necessarily cheaper either. The pros and cons _in our situation, not based on preference_ must be weighed.
```